### PR TITLE
Emit warning when using %{foo => bar} in typespecs

### PR DESCRIPTION
**Pull request by whatyouhide**
Related to #4617.

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -765,11 +765,11 @@ defmodule Kernel.Typespec do
...
         {k, v} ->
-          # TODO: Emit warnings on v1.6 (when we drop OTP 18 support)
-          # :elixir_errors.warn(caller.line, caller.file,
-          #   "invalid map specification. %{foo => bar} is deprecated in favor of " <>
-          #   "%{required(foo) => bar} and %{optional(foo) => bar}. required/1 is an " <>
-          #   "OTP 19 only feature, if you are targeting OTP 18 use optional/1.")
+          warning =
+            "invalid map specification. %{foo => bar} is deprecated in favor of " <>
+            "%{required(foo) => bar} and %{optional(foo) => bar}. required/1 is an " <>
+            "OTP 19 only feature, if you are targeting OTP 18 use optional/1."
```
Comment (whatyouhide) [inline lib/elixir/lib/kernel/typespec.ex]: @josevalim I'm not sure we should mention "if you are targeting OTP 18 use optional/1" given if you get this warning then you're on 1.6 and that does not support 18.

**Comment by josevalim** [comment]
Yes, we can remove the mention to OTP 18.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by whatyouhide** [comment]
@josevalim updated 👍

**Comment by josevalim** [comment]
We need to update the lib/elixir/pages/Deprecations.md when adding new deprecations.

**Comment by whatyouhide** [comment]
@josevalim right, updated :)

File: lib/elixir/pages/Deprecations.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ Elixir deprecations happen in 3 steps:
 
 Deprecated feature                               | Deprecated in | Replaced by (available since)
 :----------------------------------------------- | :------------ | :----------------------------
+`%{key => value}` in typespecs                   | [v1.6]        | `%{required(key) => value}` or `%{optional(key) => value}` (OTP 19)
```
Comment (josevalim) [inline lib/elixir/pages/Deprecations.md]: I think we need to list the first Elixir version that allows such syntax. I believe it is v1.4 (and we need to say it requires OTP 19 as well).

Comment (whatyouhide) [inline lib/elixir/pages/Deprecations.md]: @josevalim good point, I updated.

**Pull request merged**